s3/src/main/scala/akka/stream/alpakka/s3/impl/S3Stream.scala

```diff
@@ -65,6 +65,24 @@ private[alpakka] final class S3Stream(credentials: AWSCredentials,
     Source.fromFuture(request(s3Location, range).flatMap(entityForSuccess).map(_.dataBytes)).flatMapConcat(identity)
   }
 
+  def listBucket(bucket: String, prefix: Option[String] = None): Source[String, NotUsed] = {
+    def listBucketCall(continuation_token: Option[String] = None): Future[ListBucketResult] =
+      signAndGetAs[ListBucketResult](HttpRequests.listBucket(bucket, region, prefix, continuation_token))
+
+    def fileSourceFromFuture(f: Future[ListBucketResult]): Source[String, NotUsed] =
+      Source
+        .fromFuture(f)
+        .flatMapConcat((res: ListBucketResult) => {
```

Comment:
The recursive calls here look like a nice pattern but it doesn't run in constant memory as expected for streams and may lead to OOM eventually. See the comment https://github.com/akka/akka/issues/22352#issuecomment-281624764 here for a potential workaround.